### PR TITLE
Null annotations for java.base/java.lang

### PR DESCRIPTION
**Pull request by jason-faust**
All public and protected members and methods
Authored against OpenJDK 11.0.1 with Eclipse IDE 2018-12

### CODE REVIEW

**Comment by jason-faust** [comment]
I'll try to decipher the build errors tomorrow if I have time.

**Comment by jason-faust** [comment]
Not sure where to go with this.

Pushing things up to Java 11 requires some plugin changes, however runs into compatibility issues with `java.xml.Bind` being removed from the JDK. But even allowing that the build fails due to the `hello` project.

**Comment by vorburger** [comment]
The Travis CI build today only builds Java 8 (just fixed it). To proceed with this, you have to contribute a matrix build which also runs it under Java 11. Contributing is very welcome here!

File: libraries/java/java/lang/Class.eea [hunk 1]
```diff
@@ -1,58 +1,55 @@
...
 cast
  (Ljava/lang/Object;)TT;
  (L0java/lang/Object;)T0T;
+forName
+ (Ljava/lang/Module;Ljava/lang/String;)Ljava/lang/Class<*>;
+ (L1java/lang/Module;L1java/lang/String;)L0java/lang/Class<*>;
 forName
  (Ljava/lang/String;)Ljava/lang/Class<*>;
- (L1java/lang/String;)L1java/lang/Class<*>;
+ (Ljava/lang/String;)L1java/lang/Class<*>;
```
Comment (J-N-K) [inline libraries/java/java/lang/Class.eea:13]: This seems wrong. `Class class = Class.forName(null);` throws an NPE, to clearly the parameter must not be null.

File: libraries/java/java/lang/Class.eea [hunk 2]
```diff
@@ -1,58 +1,55 @@
...
- ()[1Ljava/lang/Class<*>;
+ ()[1L1java/lang/Class<*>;
 getComponentType
  ()Ljava/lang/Class<*>;
- ()Ljava/lang/Class<*>;
+ ()L0java/lang/Class<*>;
 getConstructor
  ([Ljava/lang/Class<*>;)Ljava/lang/reflect/Constructor<TT;>;
- ([1Ljava/lang/Class<*>;)L1java/lang/reflect/Constructor<TT;>;
+ ([0L0java/lang/Class<*>;)L1java/lang/reflect/Constructor<TT;>;
```
Comment (J-N-K) [inline libraries/java/java/lang/Class.eea:46]:
```suggestion
 ([1L1java/lang/Class<*>;)L1java/lang/reflect/Constructor<TT;>;
```

Calling `getConstructor´ passes it's `parameters` to `getConstructor0` which does something like `for (Class p : parameter)` which throws an NPE if `parameters` is `null`

File: libraries/java/java/lang/Class.eea [hunk 3]
```diff
@@ -64,7 +61,7 @@ getDeclaredClasses
  ()[1L1java/lang/Class<*>;
 getDeclaredConstructor
  ([Ljava/lang/Class<*>;)Ljava/lang/reflect/Constructor<TT;>;
- ([1Ljava/lang/Class<*>;)L1java/lang/reflect/Constructor<TT;>;
+ ([0L0java/lang/Class<*>;)L1java/lang/reflect/Constructor<TT;>;
```
Comment (J-N-K) [inline libraries/java/java/lang/Class.eea:64]:
```suggestion
 ([1L1java/lang/Class<*>;)L1java/lang/reflect/Constructor<TT;>;
```

see above

**Review by J-N-K — CHANGES_REQUESTED** [review]
I did only  check A-E yet, hope to continue tomorrow.

File: libraries/java/java/lang/Process.eea [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ ()L1java/util/stream/Stream<L1java/lang/ProcessHandle;>;
+descendants
+ ()Ljava/util/stream/Stream<Ljava/lang/ProcessHandle;>;
+ ()L1java/util/stream/Stream<L1java/lang/ProcessHandle;>;
+destroyForcibly
+ ()Ljava/lang/Process;
+ ()L1java/lang/Process;
+getErrorStream
+ ()Ljava/io/InputStream;
+ ()L1java/io/InputStream;
```
Comment (J-N-K) [inline libraries/java/java/lang/Process.eea:13]:
```suggestion
 ()L0java/io/InputStream;
```

in some cases a null input stream is returned

**Review by J-N-K — CHANGES_REQUESTED** [review]
How shall we proceed now?

File: libraries/java/java/lang/Class.eea (hunk 1 above)
Comment (jason-faust) [reply libraries/java/java/lang/Class.eea:13]: You are correct, fixed.

File: libraries/java/java/lang/Class.eea (hunk 2 above)
Comment (jason-faust) [reply libraries/java/java/lang/Class.eea:46]: I'm not seeing this being thrown in my testing, but I'm also looking at this with Java 14 at the moment.

Comment (jason-faust) [reply libraries/java/java/lang/Class.eea:46]: Pulled down a copy of the Class source file for JDK 11, code path looks the same, the parameters get looked at in `arrayContentsEq()`, and that has a proper null test.

The below works for me.
```java
package npa;
import java.util.ArrayList;
public class Test {
	public static void main(String[] args) throws Exception {
		Class<ArrayList> a = ArrayList.class;
		System.out.println(a.getConstructor(null));
		System.out.println(a.getDeclaredConstructor(null));
	}
}
```

File: libraries/java/java/lang/Process.eea [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+getInputStream
+ ()Ljava/io/InputStream;
+ ()L1java/io/InputStream;
```
[the remark this answers is not in the document]
Comment (jason-faust) [reply libraries/java/java/lang/Process.eea:16]: This doesn't return a null value, it returns a null stream, which is still a valid object reference.  See https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/lang/ProcessBuilder.html#redirect-output

[the remark this answers is not in the document]
Comment (J-N-K) [reply libraries/java/java/lang/Process.eea:16]: My fault.

File: libraries/java/java/lang/Class.eea (hunk 2 above)
Comment (J-N-K) [reply libraries/java/java/lang/Class.eea:46]: I‘ll re-check that. Pretty sure I had a test-case where I encountered an NPE.

**Review by J-N-K — APPROVED** [review]

**Comment by J-N-K** [comment]
I guess Travis fails because the methods are missing in Java 8.

File: libraries/java/java/lang/Process.eea (hunk 5 above)
[the remark this answers is not in the document]
Comment (jason-faust) [reply libraries/java/java/lang/Process.eea:16]: Don't think there is anything to resolve, the annotations show the return value is non null.  The stream is useless but is valid as far as nulls are conserned.

**Comment by jason-faust** [comment]
@J-N-K from what @vorburger said, the build needs to be re-configured to support JDK 8 and 11 (and 14), which is impossible with just a simple switch to the matrix build (I tried).

**Comment by J-N-K** [comment]
The problem will be that we would have to add additional annotations so the tests know what the correct java version for this annotation is, otherwse the Java8 build will always fails.

Another solution would be to have two branches (and two releases), one for Java8 and one for Java 11. This would require cherry-picking compatible changes from the 11-branch to the 8-branch.

The easiest would be to change to Java11 now, since Java8 is now six years old and EOL (at least for Oracle versions),

**Comment by J-N-K** [comment]
@vorburger @kaikreuzer How can we proceed here? I would suggest to change Travis from Java 8 to Java 11 and merge here. A new release afterwards would be great.

**Comment by vorburger** [comment]
> @vorburger @kaikreuzer How can we proceed here? I would suggest to change Travis from Java 8 to Java 11 and merge here. A new release afterwards would be great.

Thank You for your interest in [LastNPE.org](http://www.lastnpe.org/)! While I currently do not have the spare cycles to continue contributing to actively maintain this community, I'm very happy to see both of your interests, and would hereby like to empower you to "carry the torch forward" - should you be interested to do so; see #91. Should you accept, feel free to change Travis - and go ahead making any other changes to this project you deem useful.

**Pull request merged**